Spec/Auth.swift

```diff
@@ -3619,5 +3619,331 @@ class Auth : QuickSpec {
+                options.autoConnect = false
+                options.authUrl = NSURL(string: echoServerAddress)! as URL
+
+
+                context("with valid credentials") {
+                    options.authParams = [URLQueryItem]() as [URLQueryItem]?
+                    options.authParams?.append(URLQueryItem(name: "keyName", value: keys["keyName"]) as URLQueryItem)
+                    options.authParams?.append(URLQueryItem(name: "keySecret", value: keys["keySecret"]) as URLQueryItem)
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+
+                    it("fetches a channels and posts a message") {
+                        client.channels.get(channelName).publish(messageName, data: nil, callback: { error in
+                            expect(error).to(beNil())
+                        })
+                    }
+                }
+
+                context("with wrong credentials") {
+                    options.authParams = [URLQueryItem]() as [URLQueryItem]?
+                    options.authParams?.append(URLQueryItem(name: "keyName", value: keys["keyName"]) as URLQueryItem)
+                    options.authParams?.append(URLQueryItem(name: "keySecret", value: "INVALID") as URLQueryItem)
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+
+                    it("fails to connect with reason 'invalid signature'") {
+                        waitUntil(timeout: testTimeout) { done in
+                            client.connection.once(.failed) { stateChange in
+                                expect(stateChange!.reason!.code).to(equal(40144))
+                                expect(stateChange!.reason!.description).to(contain("invalid signature"))
+                                done()
+                            }
+                            client.connect()
+                        }
+                    }
+                }
+
+                context("when token expires") {
+                    let tokenDuration = 5.0
+                    options.authParams = [URLQueryItem]() as [URLQueryItem]?
+                    options.authParams?.append(URLQueryItem(name: "keyName", value: keys["keyName"]) as URLQueryItem)
+                    options.authParams?.append(URLQueryItem(name: "keySecret", value: keys["keySecret"]) as URLQueryItem)
+                    options.authParams?.append(URLQueryItem(name: "expiresIn", value: String(UInt(tokenDuration))) as URLQueryItem)
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+
+                    it ("receives a 40142 error from the server") {
+                        waitUntil(timeout: testTimeout) { done in
+                            client.connection.once(.connected) { stateChange in
+                                delay(tokenDuration + 1) {
```

Comment:
Why do you need this delay?  Should the connection `once(.disconnected)` fire below anyway once the token has expired?  In fact, if waits `tokenDuration`, what's probably happening with this test is that it's being disconnected once here, and once again below.

Reply:
Done https://github.com/ably/ably-ios/pull/714/commits/936955b51ed438eb74ed0f0f5036e95f9cbc0a22